Add support for Visual Studio Team Services and Team Foundation Server

ethomson opened the pull request:
Microsoft [Visual Studio Team Services](https://www.visualstudio.com/team-services/) is a cloud-hosted DevOps platform that provides Git repository hosting.  [Team Foundation Server](https://www.visualstudio.com/tfs/) is the on-premises version of VSTS.  I've added some changes to support them:

1. Determine whether a remote path is a URL or an SCP style path, and parse them separately.  This allows us to understand the URL scheme, to know whether we should preserve HTTP semantics and port information:
      * VSTS advertises SSH paths with a port, which should _not_ be preserved in the resultant web URL
      * TFS can use HTTP paths when on-premises, which should be preserved
2. Branch information for VSTS and TFS is suffixed to the path using a query string and not additional path components.  I've changed the other implementations `providerBranchRef` to include the `/` prefix themselves, so that we do not join the `urlpath` and `providerBranchRef` on VSTS/TFS using a `/`.  (This would result in a URL that was `https://my.visualstudio.com/_git/repository/?branch=...`.  The slash before the query string is incorrect in VSTS URLs.)
3. Add support for issues ("workitems" in VSTS parlance).

ethomson commented:
BTW - I really love the unit tests.  I haven't really used bats before in anger - in a previous job, we relied pretty heavily on a bespoke bourne shell testing framework.  And of course git itself has _its_ own bespoke bourne shell testing framework.  In any case, I'm very glad to see the unit tests.  👍

derimagia commented:
Do you mind taking a look at https://github.com/paulirish/git-open/pull/92 and seeing if the parsing is easier after that change? That should make the first commit a lot simpler.

ethomson commented:
I didn't see #92 when I created this.  I'm happy to rebase on top of that if it gets merged.  @paulirish ?

erendrake commented:
the ssh urls for VSTS are about to change 

```remote:
remote: ********************************************************************
remote: *       After November 17th, 2017.                                 *
remote: *       Old SSH remote URLs will be retired and                    *
remote: *       you will need to update your SSH Git remotes.              *
remote: *                                                                  *
remote: *       If you haven't already done this:                          *
remote: *       - Visit your VSTS repository on the web.                   *
remote: *       - Press the "Clone" button on the top right corner.        *
remote: *       - Press "SSH" and copy the new SSH remote URL.             *
remote: *       - In your Git client, run the following:                   *
remote: *             git remote set-url {remote_name} {new_ssh_url}       *
remote: *                                                                  *
remote: *       For more information, refer to                             *
remote: *             https://go.microsoft.com/fwlink/?linkid=855768       *
remote: ********************************************************************
remote:```

derimagia commented:
Can you check if #92 fixes this for you? It'll allow you to change the domain. I'll update the readme with it and merge it in if it works for you.

ethomson commented:
Manually changing the domain seems orthogonal to this problem, no?

derimagia commented:
Sorry you're right, the config was mainly for hosted domains. The one thing that PR does do is make some of this code unneeded since it should properly parse scp urls. When I get a chance I'll do another 1-over on the PR and once it's merged if you can rebase it and if it can be updated with the new domain I think this PR will be good to merge too

erendrake commented:
sorry if my previous post was confusing. you are right that manual updates are not important to this PR. I was mostly trying to point out that one of your tests has a url that will not be correct soon. 

namely:

[https://github.com/ethomson/git-open/blob/vsts/test/git-open.bats#L260](this one)

it currently reads 

ssh://gitopen@gitopen.visualstudio.com:22/Project/_git/Repository

after the change its ssh url will change to 

ssh://gitopen@**vs-ssh**.visualstudio.com:22/Project/_git/Repository

ethomson commented:
Thanks @erendrake.  I had mostly forgotten about this pr.  Will update with new ssh urls pointing at the edge Azure termination endpoints.

paulirish commented:
@ethomson i've merged #92 so you can now rebase on top.  I'm hoping the work in 92 will simplify some of the changes you had to make here.  G'luck! :)

ethomson commented:
I've rebased this.

Note that I have reverted a change introduced in #92 that tests that port numbers are _removed_ from HTTPS.  This affects the case where your git endpoint is `https://foo:4444/project.git`:

I've installed GitLab CE so that I could change the port and confirm that _both_ the Git endpoint _and_ the actual web endpoint (ie, repository browsing) are affected.  So reverting this does seem like the correct behavior in general.  I would suggest that if people have some non-standard split setup that it is the much less common configuration, and they should use the new configuration options to `git-open` to fix that.  But perhaps I'm misunderstanding something.

ethomson commented:
@erendrake I have not updated the behavior to reflect the new SSH URL scheme for VSTS.  At the moment, there's no mapping between SSH URLs and your HTTP endpoint.  I'll see if we can add a wildcard DNS mapping for `vs-ssh.visualstudio.com` and then recommend people use `tenant.vs-ssh.visualstudio.com` so that we can map that back to `tenant.visualstudio.com`.  But I don't think that this issue should preclude moving forward with general VSTS and TFS support.

derimagia commented on git-open:
```diff
@@ -97,30 +97,34 @@ IFS='/' pathargs=($urlpath)
   IFS='/' urlpath="${pathargs[*]}"
-  providerBranchRef="browse?at=$branch"
+  providerBranchRef="/browse?at=$branch"
+elif [[ "${#pathargs[@]}" -ge "2" && ${pathargs[${#pathargs[@]} - 2]} == "_git" ]]; then
```
@paulirish do you know if shellcheck supports linting like consistent quoting? I don't really care too much if we use single/double quotes but I used single quotes elsewhere

paulirish replied:
poking around I'm not seeing much. The closest issue i can find was in https://github.com/koalaman/shellcheck/issues/689 but that was closed in shellcheck 0.4.6 which I just confirmed is the version running on travis.

Additionally it appears that all shellcheck checks are run by default and are just disabled individually... so, no additional rules we even could enable. ¯\_(ツ)_/¯

ethomson replied:
Fixed this to be single-quotes for consistency.

paulirish commented on git-open:
```diff
@@ -46,17 +46,37 @@ fi
-domain=${uri%%[:$pathsep]*} urlpath=${uri#*$pathsep}
+# If the URL is provided as "http", preserve that, otherwise assume "https"
+if [[ $protocol != "http" ]]; then
+  protocol="https"
```
since we already have https as the default... seems like we shouldn't have to reset it back to that default value.

perhaps in your diff you could set a `gitprotocol` variable .. then here you could just test that and change `protocol` to `"http"` if necessary.

ethomson replied:
Set a `gitprotocol` variable and update `protocol` iff `gitprotocol == 'https'`.

paulirish commented on git-open:
```diff
@@ -46,17 +46,37 @@ fi
+    domain=${domain%:*}
+  fi
+else
+  # Trip possible username from SSH path
```
Trim

ethomson replied:
Fixed, thanks.

paulirish commented on git-open:
```diff
@@ -77,30 +97,40 @@ IFS='/' pathargs=($urlpath)
-  providerBranchRef="browse?at=$branch"
+  providerBranchRef="/browse?at=$branch"
+elif [[ "${#pathargs[@]}" -ge "2" && ${pathargs[${#pathargs[@]} - 2]} == "_git" ]]; then
+  # Visual Studio Team Services and Team Foundation Server
```
can you add a comment saying VSTS always will have /_git/ as the second to last segment in the url path?

a bit easier to read than parsing L117 ;)

ethomson replied:
Definitely.  Added a comment to clarify what was going on here.

paulirish commented:
Great tests. thanks for the rebase, too. :)

ethomson commented:
Fixed up and rebased.  Thanks!

ethomson commented on README.md:
````diff
@@ -12,7 +12,7 @@ git open [remote-name] [branch-name]
 ```
 
-(`git open` works with these [hosted repo providers](#supported-remote-repositories), `git open --issue` currently only works with GitHub)
+(`git open` works with these [hosted repo providers](#supported-remote-repositories), `git open issue` currently only works with GitHub, Visual Studio Team Services and Team Foundation Server)
````
Right you are; fixed.  Thanks!

paulirish commented:
Thanks very much @ethomson!

ethomson commented:
Thanks everybody!  ✨ 

Curious when the next release might be?

paulirish commented:
We're working towards shipping a 2.0: https://github.com/paulirish/git-open/milestone/1

and we're just about done....
